### ISSUE TITLE
Some URLs cause Segfault via std::regex exception not caught

### ISSUE THREAD

**Issue by beneschtech** [etr/libhttpserver#196]
### Prerequisites

* [X ] Put an X between the brackets on this line if you have checked that your issue isn't already filed: https://github.com/search?l=&q=repo%3Aetr%2Flibhttpserver&type=Issues

### Description

Issuing a url to your service with { or [ or its closing counterparts causes a segfault via an uncaught std::regex exception.

### Steps to Reproduce

1. Write any libhttpserver based service, even the example helloworld exhibits this behavior
2. Issue something like the following curl 'http://localhost:<port>/endpoint/\{\}'

This was caused by a naughty application talking to it, and has been fixed on its side, but it does leave a DOS hole for an attacker to shut down your service.

**Expected behavior:** Anything besides segfault?  Maybe a 500 error?

**Actual behavior:** Program terminates due to uncaught std::regex exception in http_endpoint

**Reproduces how often:** Every time

### Versions

* OS version (if on linux, the output of "uname -a")  CentOS 7,8, Fedora 31
* libhttpserver version (please specify whether compiled or packaged) 0.18 compiled
* libmicrohttpd version (please specify whether compiled or packaged) 0.9.70,0.9.59 - packaged

If you have problems during build:
* Compiler version
* autotools version

### Additional Information

Any additional information, configuration (especially build configuration flags if you compiled the libraries) or data that might be necessary to reproduce the issue.

If you have problems during build, please attach your config.log and the full scope of your error from make.

If you have problems at execution, please:
* attach the stacktrace in case of crash (a coredump would be even better).
* provide a main that reproduces the error.

I've been a C++ developer for 20 years, I would be glad to fix this.

**Comment by etr** [comment]
Thanks - this is a pretty bad bug. I am really sorry that you had to encounter this.

I just added a fix in a branch.

One way to prevent it, before my change goes to master which is at latest tomorrow, is to disable the regex matching entirely (at least you are de-risked).

Once this is on master and built by travis, I will make a release too.

This is actually due to the fact that the regex pattern is built regardless of whether the http_endpoint class is used for the registration of a resource or for the matching of the same.

**Comment by beneschtech** [comment]
No problem, it took me forever to track down why my server process kept
dieing.
This is a great piece of software, good job guys!
I think the regex might be a little faster, but just in case, I didnt
catch where in the documentation I could turn it off.  I use urls like
"/aoi" with family true, but even on family set to false, ie "/stop" it
does it.

**Comment by etr** [comment]
You can disable regexes on your server by setting ".no_regex_checking()" in the create_webserver class. This disables regex checking entirely on your server. If you don't use them anyhow, it just makes the webserver way faster.

**Comment by etr** [comment]
The issue is fixed on head and in release 0.18.1. Thanks again for notifying.

Feel free to re-open in case there is still a problem on your side.

**Comment by beneschtech** [comment]
Worked like a charm, thank you!
Next weekend I'll look at the bug list and see what I can help clean
up.  I'd love to contribute to this great project.
Example:[chris@fbsd-wkstn1 /home/chris/Beneschtech/pubapi-
web/src/apiserver] > curl -k -i -v 'https://localhost:8192/ping/\{\}'* 
  Trying ::1:8192...* Connected to localhost (::1) port 8192 (#0)*
ALPN, offering h2* ALPN, offering http/1.1* successfully set
certificate verify locations:*   CAfile: /usr/local/share/certs/ca-
root-nss.crt  CApath: none* TLSv1.3 (OUT), TLS handshake, Client hello
(1):* TLSv1.3 (IN), TLS handshake, Server hello (2):* TLSv1.3 (IN), TLS
handshake, Encrypted Extensions (8):* TLSv1.3 (IN), TLS handshake,
Certificate (11):* TLSv1.3 (IN), TLS handshake, CERT verify (15):*
TLSv1.3 (IN), TLS handshake, Finished (20):* TLSv1.3 (OUT), TLS change
cipher, Change cipher spec (1):* TLSv1.3 (OUT), TLS handshake, Finished
(20):* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384* ALPN,
server did not agree to a protocol* Server certificate:*  subject:
C=US; ST=Colorado; L=Colorado Springs; O=Beneschtech, LLC;
OU=Development; CN=beneschtech.com;
emailAddress=chris@beneschtech.com*  start date: May 25 10:52:30 2020
GMT*  expire date: May 25 10:52:30 2021 GMT*  issuer: C=US;
ST=Colorado; L=Colorado Springs; O=Beneschtech, LLC; OU=Development;
CN=beneschtech.com; emailAddress=chris@beneschtech.com*  SSL
certificate verify result: self signed certificate (18), continuing
anyway.> GET /ping/{} HTTP/1.1> Host: localhost:8192> User-Agent:
curl/7.69.1> Accept: */*> * Mark bundle as not supporting multiuse<
HTTP/1.1 404 Not FoundHTTP/1.1 404 Not Found< Connection: Keep-
AliveConnection: Keep-Alive< Content-Length: 9Content-Length: 9<
Content-Type: text/plainContent-Type: text/plain< Server:
Beneschtech/APIServer 1.0Server: Beneschtech/APIServer 1.0< Date: Sat,
06 Jun 2020 19:45:13 GMTDate: Sat, 06 Jun 2020 19:45:13 GMT
< * Connection #0 to host localhost left intact

**Comment by beneschtech** [comment]
Hey Sebastiano,
I just discovered that family doesnt work with disabled regex.  Ie I
register a resource /aoi with family set to true, and the webserver set
to no_regex_checking, then when I make a request of the sort /aoi/CON I
get a 404 error.
If it was by design thats cool, I was just curious.

**Comment by etr** [comment]
Sorry, I did not realize that you needed family URLs. By the way, the current solution on head, and latest release fix the problem in general.

For the family and regex, that is by-design or, at least, it made sense when it was built. Originally "family" was a short/quick way to do what is most commonly done whenever regexes were used. In all honesty, I don't think that is evident in the documentation.

Admittedly, if only a prefix is needed, we could make that much cheaper by having just the prefix being checked. Decoupling the two might be an interesting feature to add.

**Comment by beneschtech** [comment]
Oh, absolutely.  I cant stress enough how useful and efficient this
software is.  I am trying to use it in a pretty high volume environment
and am slowly consolidating a very old, very big software package that
relied on sun RPC for literally decades to have a centralized endpoint
for its services that can also be used by customers if they want and am
doing my best to adhere to traditional REST protocol.  Theres always
single resource, but I think a lot of users may not realize the
performance penalty of regex, useful as it may be.
My thought was a linked list with the url prefix, and a hit counter
(atomic unsigned long) using a passive bubble sort.  That way more
frequently hit urls would rise to the top and be the first to be
checked.  That is of course after the inital std::map find if it isnt
successfull, and regex is off.
Again, I would be more than happy to do it myself to contribute.  I can
do some code cleanup too, ie const bool function() const throws a
warning in FreeBSD and Linux.

**Comment by etr** [comment]
Hey, thanks for the nice feedback.

Your use-case is close to the one that I found myself into originally when I wrote this.

If everything goes to a single resource, you might also want to take a look at .single_resource() in the webserver builder class. That allows to register a single resource that will receive every request regardless of the path (effectively if bypasses routing entirely). Not sure it applies, but worth mentioning.

On your offer to help, I'd be more than glad to support the feature of having "family" resources not paying the cost of regex whenever it is not needed and I will be happy to review any PR that you send over.
